add medallia styling as separate output

rianfowler opened the pull request:
## GitHub Issue

[Host Medallia CSS #13287](https://github.com/department-of-veterans-affairs/va.gov-team/issues/13287) 

## Description

Remove hash from Medallia CSS file in staging environment

## Testing done

1. Run `yarn build:webpack:prod`
   - Build time with sourcemaps = ~50 minutes 😱 
   - Build time without sourcemaps = ~15 minutes 😐 
1. Run `ls -l build/vagovprod/generated | grep medal` to find the medallia CSS file
1. Confirm there is a hash in the filename (for example, `va-medallia-styles.a9a525ad4ac2474ed8b4-1601477867140.css`)
1. Run `yarn build:webpack:staging`
1. Run `ls -l build/vagovstaging/generated | grep medal` to find the medallia CSS file
1. Confirm there is no hash in the filename (for example, `va-medallia-styles.css`)

## Screenshots

![](https://user-images.githubusercontent.com/6130520/94706143-71924d80-0307-11eb-9b44-ef78f3f3b425.png)

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

Commit:
add medallia styling as separate output

Commit:
fix reference to va-medallia-style

Commit:
update filenaming logic for css to exclude medallia styles from filename hashing

Commit:
merge master

billfienberg commented on config/webpack.config.js:
```diff
@@ -236,9 +241,14 @@ module.exports = env => {
-        filename: !isOptimizedBuild
-          ? '[name].css'
-          : `[name].[contenthash]-${timestamp}.css`,
+        moduleFilename: ({ name, entryModule }) => {
```
`moduleFilename` docs: https://webpack.js.org/plugins/mini-css-extract-plugin/#module-filename-option

billfienberg replied:
Here are the keys that are on the object argument for the `moduleFilename` function:
```
[ 'id',
  'ids',
  'debugId',
  'name',
  'preventIntegration',
  'entryModule',
  '_modules',
  'filenameTemplate',
  '_groups',
  'files',
  'rendered',
  'hash',
  'contentHash',
  'renderedHash',
  'chunkReason',
  'extraAsync',
  'removedModules' 
]
```

billfienberg replied:
https://webpack.js.org/configuration/output/#template-strings

![image](https://user-images.githubusercontent.com/6130520/94280978-ae2f0500-ff13-11ea-9d1c-8ec9f736fc70.png)

billfienberg committed:
Merge branch 'master' into add-medallia-styling

billfienberg committed:
package.json: Add build:webpack:prod npm script

billfienberg commented:
Copying a Slack conversation here for visibility
>**Rian**: tl;dr we need to push just this css file out without a hash on the filename. I have figured out how to configure webpack to do that but the output is not matching the master branch output.
>
>once we can publish the file to staging, we need to give it to insights team to add to the medallia iframe settings. it will load the css dynamically and we may need to do some tweaks to make the survey modal look like the design.
>
>**Bill**: Which output? The entire `build` folder? The `generated` folder? The medalia specific file?
>
>**Rian**: the generated folder
>sorry i should have been more specific about this
>i’ve been running this command: `NODE_ENV=production yarn build:webpack --env.buildtype vagovprod`
>on my branch vs master branch and comparing the output of build/vagovprod
>that process takes forever which is the hard part 
>that argument for the `moduleFileName` is not documented anywhere- there are 3 different hashes with different names in that available as an argument and i may not be using the right one
>
>**Bill**: I suspected it was `generated`, but I've been diffing the entire build to be safe. And those builds are slow.
>
>**Rian**: what i saw was that my branch build output more files than the master branch- more things without hashes which was confusing
>yeah and you have to make sure your master / local branches are on the same commit
>i would expect the hashes to be the same but i may not be using the same hash `renderedHash` may not be what `filename` is using
>so you might want to console log out that object to see what all of the options on i should have captured it somewhere

billfienberg commented on package.json:
```diff
@@ -11,6 +11,7 @@
     "build:content": "node --max-old-space-size=8192 script/build-content.js",
     "build:content:test": "node script/test-cms-export.js",
     "build:webpack": "NODE_OPTIONS=--max-old-space-size=4096 webpack --config config/webpack.config.js",
+    "build:webpack:prod": "NODE_ENV=production yarn build:webpack --env.buildtype vagovprod",
```
This is intended to be temporary. I just added it so I don't mistakenly run `yarn build` and end up with a `build/localhost` folder.

rianfowler replied:
looks good

billfienberg commented:
Disabling source maps reduced my build from ~3,000 seconds (~50 minutes 😱 ) to 625.29s. (~10.5 minutes 😐 ).

billfienberg committed:
webpack.config.js: Temporarily disable sourcemaps

billfienberg commented on config/webpack.config.js:
```diff
@@ -13,14 +13,16 @@ const ManifestPlugin = require('webpack-manifest-plugin');
 
 const headerFooterData = require('../src/platform/landing-pages/header-footer-data.json');
-const BUCKETS = require('../src/site/constants/buckets');
+// const BUCKETS = require('../src/site/constants/buckets');
```
Temporary - will be deleted before merging

billfienberg commented on config/webpack.config.js:
```diff
@@ -148,7 +153,7 @@ module.exports = env => {
             {
               loader: 'sass-loader',
-              options: { sourceMap: true },
+              // options: { sourceMap: true },
```
Temporary - will be deleted before merging

billfienberg commented on config/webpack.config.js:
```diff
@@ -214,7 +219,7 @@ module.exports = env => {
           // cache: true,
           parallel: 3,
-          sourceMap: true,
+          // sourceMap: true,
```
Temporary - will be deleted before merging

billfienberg commented on config/webpack.config.js:
```diff
@@ -373,14 +383,14 @@ module.exports = env => {
-        filename: '[file].map',
-      }),
-    );
+    // const bucket = BUCKETS[buildOptions.buildtype];
```
Temporary - will be deleted before merging

billfienberg commented on config/webpack.config.js:
```diff
@@ -373,14 +383,14 @@ module.exports = env => {
-    );
+    // const bucket = BUCKETS[buildOptions.buildtype];
+
+    // baseConfig.plugins.push(
```
Temporary - will be deleted before merging

billfienberg commented:
I wonder if we should be using `contentHash` instead of `renderedHash`. 

>TLDR: Use [contenthash] instead of [chunkhash] to ensure that every asset (output file) has a stable hash based on it's contents (only). - https://github.com/webpack/webpack/issues/7179#issuecomment-386132702

billfienberg commented:
I wonder if we should try adding `optimization: moduleIds: 'hashed'`:

>Now, despite any new local dependencies, our vendor hash should stay consistent between builds: - https://webpack.js.org/guides/caching/#module-identifiers

>`moduleIds: 'deterministic'` was added in webpack 5 and `moduleIds: 'hashed'` is deprecated in favor of it. - https://webpack.js.org/configuration/optimization/#optimizationmoduleids

billfienberg commented:
I'm seeing lots of `*.undefined-${timestamp}.css` when using `${contenthash}`. 🤔 

![image](https://user-images.githubusercontent.com/6130520/94481690-bef2ab80-019d-11eb-844e-586cd702c958.png)

billfienberg commented:
🤦‍♂️ 

The MiniCssExtractPlugin plugin options seem to be documented in a JSON file. 

https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/src/plugin-options.json

billfienberg committed:
package.json: Add debug:webpack:prod npm script

billfienberg committed:
webpack.config.js: Use 'contenthash' substitution instead of interpolated 'entryModule.renderedHash'

billfienberg committed:
Revert "webpack.config.js: Temporarily disable sourcemaps"

This reverts commit 646a5e67b81055704a99b6eb5c06d5ba227a4e46.

billfienberg commented on package.json:
```diff
@@ -11,6 +11,8 @@
     "build:content:test": "node script/test-cms-export.js",
     "build:webpack": "NODE_OPTIONS=--max-old-space-size=4096 webpack --config config/webpack.config.js",
+    "build:webpack:prod": "NODE_ENV=production yarn build:webpack --env.buildtype vagovprod",
+    "debug:webpack:prod": "NODE_ENV=production NODE_OPTIONS=--max-old-space-size=4096 node --inspect --inspect-brk ./node_modules/webpack/bin/webpack.js --config config/webpack.config.js --env.buildtype vagovprod",
```
1. Run `debug:webpack:prod`
1. Open chrome://inspect
1. Click **Inspect** for the remote target (./node_modules/webpack/bin/webpack.js)

![debug-node](https://user-images.githubusercontent.com/6130520/94571153-463f2e00-0235-11eb-9c68-97fb6da33b48.gif)

Reference: https://webpack.js.org/contribute/debugging/

billfienberg commented on config/webpack.config.js:
```diff
@@ -236,9 +241,14 @@ module.exports = env => {
-        filename: !isOptimizedBuild
-          ? '[name].css'
-          : `[name].[contenthash]-${timestamp}.css`,
+        moduleFilename: chunk => {
```
https://github.com/webpack-contrib/mini-css-extract-plugin/issues/67#issuecomment-520795293

billfienberg committed:
Merge branch 'master' into add-medallia-styling

billfienberg commented:
Related tickets:

- [Upgrade to Webpack 5](https://github.com/department-of-veterans-affairs/va.gov-team/issues/11510) 
- [Modularize the Webpack config](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2796)

billfienberg committed:
webpack.config.js: Remove hash from Medallia style filename in staging

billfienberg committed:
package.json: Add build:webpack:staging, build:webpack:dev, and build:webpack:local npm scripts

billfienberg commented:
`build/vagovprod` = `va-medallia-styles.a9a525ad4ac2474ed8b4-1601477867140.css`
`build/vagovstaging` = `va-medallia-styles.css`

![image](https://user-images.githubusercontent.com/6130520/94706143-71924d80-0307-11eb-9b44-ef78f3f3b425.png)

billfienberg committed:
webpack.config.js: Update Medallia spelling (from medalia to medallia)

U-DON commented on config/webpack.config.js:
```diff
@@ -236,9 +241,20 @@ module.exports = env => {
+          const isStaging =
+            buildOptions.buildtype === ENVIRONMENTS.VAGOVSTAGING;
+
+          return isStaging ? `${name}.css` : standardFilename;
```
Is this for testing on staging first?

billfienberg replied:
That's correct. 

>once we can publish the file to staging, we need to give it to insights team to add to the medallia iframe settings. - https://github.com/department-of-veterans-affairs/vets-website/pull/14312#issuecomment-700067751

billfienberg replied:
>I think we should have the staging check before any return statements. otherwise this logic is fine.

@rianfowler, I don't have an objection to moving the staging check before any `return`s. That being said, is the Medallia CSS file the only file that cares about the staging check? If so, what is the value of performing the staging check for every file?

billfienberg replied:
>Now that I'm getting a second look at this, it seems like dev would keep the hash in the name (not optimized build, so showHash is falsy). Not sure how much this matters.

@U-DON, my intent for the code was:
1. Determine if it's the Medallia CSS file.
1. If not, return our normal filename format.
1. If so, determine if it's the staging* environment.
   * My understanding of the requirements is that I should be checking only for staging. If I should be checking for additional non-prod environments, then I need to update my code. 
1. If staging, exclude the hash. 
1. If non-staging, include the hash.

U-DON commented on config/webpack.config.js:
```diff
@@ -21,6 +21,8 @@ const {
   getWebpackEntryPoints,
 } = require('./manifest-helpers');
 
+const vaMedaliaStylesFilename = 'va-medallia-styles';
```
```suggestion
const vaMedalliaStylesFilename = 'va-medallia-styles';
```

billfienberg replied:
🤦‍♂️  I made that change locally earlier today, but forgot to push it. 

0207819b6edddcb09f8f52f2b5f2902d7f9c15e1

rianfowler commented on config/webpack.config.js:
```diff
@@ -21,6 +21,8 @@ const {
   getWebpackEntryPoints,
 } = require('./manifest-helpers');
 
+const vaMedalliaStylesFilename = 'va-medallia-styles';
```
```suggestion
// TODO: refactor the other approach for creating files without the hash so that we're only doing that in the webpack config: https://github.com/department-of-veterans-affairs/vets-website/blob/a012bad17e5bf024b0ea7326a72ae6a737e349ec/src/site/stages/build/plugins/process-entry-names.js#L35
const vaMedalliaStylesFilename = 'va-medallia-styles';
```

billfienberg committed:
Update config/webpack.config.js

Co-authored-by: Rian Fowler <rian.fowler@adhocteam.us>

billfienberg committed:
webpack.config.js: Replace temp 'isStaging' variable with 'isStaging' environment helper

billfienberg commented on config/webpack.config.js:
```diff
@@ -236,9 +241,20 @@ module.exports = env => {
+
+          if (showHash) return standardFilename;
+
+          const isStaging =
```
replaced temp variable with environment helper in 0412326a8c5d81c670da44f33c43c7261e83ff1a

billfienberg replied:
I went back to using a temp variable instead of the environment helper function in f832dd3f7bd09395c86b9ea15a1b81f555e7cdb2. 

The [build was failing](https://app.circleci.com/pipelines/github/department-of-veterans-affairs/vets-website/14370/workflows/7637a22e-c24d-480d-b719-c20fdf082156/jobs/110434/parallel-runs/0/steps/0-105). I suspect it's because the helper function uses an `import` statement, and we're using `node@10`, which doesn't support `import` statements. 
```
import ENVIRONMENTS from 'site/constants/environments';
       ^^^^^^^^^^^^
SyntaxError: Unexpected identifier
```

![image](https://user-images.githubusercontent.com/6130520/94846819-60207280-03e7-11eb-952d-0a10dcc27fcc.png)

U-DON commented on config/webpack.config.js:
```diff
@@ -236,9 +243,17 @@ module.exports = env => {
-          : `[name].[contenthash]-${timestamp}.css`,
+        moduleFilename: chunk => {
+          const { name } = chunk;
+          const standardFilename = `${name}.[contenthash]-${timestamp}.css`;
+          const isMedalliaStyleFile = name === vaMedalliaStylesFilename;
+          const isStandardFile = !isMedalliaStyleFile;
+          const showHash = isOptimizedBuild && isStandardFile;
+
+          if (showHash) return standardFilename;
+
+          return environment.isStaging() ? `${name}.css` : standardFilename;
```
>> Now that I'm getting a second look at this, it seems like dev would keep the hash in the name (not optimized build, so showHash is falsy). Not sure how much this matters.
> 
> @U-DON, my intent for the code was:
> 
> 1. Determine if it's the Medallia CSS file.
> 1. If not, return our normal filename format.
> 1. If so, determine if it's the staging* environment.
>     - My understanding of the requirements is that I should be checking only for staging. If I should be checking for additional non-prod environments, then I need to update my code.
> 1. If staging, exclude the hash.
> 1. If non-staging, include the hash.

Restarting thread on new changes so the discussion doesn't get lost from an outdated snippet.

If the requirements is only for staging, that's fair. I just wanted to point out that this looks like it woud add hashes for dev, which is a deviation from the original code.

If we want to be explicit about "hash only for Medallia, only in staging", then would it be clearer to write it like this?

```suggestion
          const standardFilename = `${name}.[contenthash]-${timestamp}.css`; 
          const isMedalliaStyleFile = name === vaMedalliaStylesFilename; 
          
          if (isMedalliaStyleFile && environment.isStaging()) return `${name}.css`;
          
          return isOptimizedBuild ? standardFilename : `${name.css}`;
```

billfienberg replied:
This looks a lot cleaner. Thanks for the suggestion.

billfienberg committed:
Update config/webpack.config.js

Co-authored-by: Eugene Doan <1067024+U-DON@users.noreply.github.com>

billfienberg committed:
webpack.config.js: Replace isStaging helper function with temp variable

SyntaxError: Unexpected identifier coming from import ENVIRONMENTS from 'site/constants/environments';
I suspect it's because we're using node@10, which doesn't support import statements.

billfienberg committed:
webpack.config.js: Use webpack substitutions instead of ES6 string interpolations

billfienberg committed:
webpack.config.js: Delete leftover closing curly brace
